x-pack/plugins/encrypted_saved_objects/server/saved_objects/encrypted_saved_objects_client_wrapper.ts

```diff
@@ -68,18 +60,21 @@ export class EncryptedSavedObjectsClientWrapper implements SavedObjectsClientCon
     }
 
     // Saved objects with encrypted attributes should have IDs that are hard to guess especially
-    // since IDs are part of the AAD used during encryption. Types can opt-out of this restriction,
-    // when necessary, but it's much safer for this wrapper to generate them.
-    if (
-      options.id &&
-      !this.options.service.canSpecifyID(type, options.version, options.overwrite)
-    ) {
+    // since IDs are part of the AAD used during encryption, that's why we control them within this
+    // wrapper and don't allow consumers to specify their own IDs directly.
+
+    // only allow a specified ID if we're overwriting an existing ESO with a Version
+    // this helps us ensure that the document really was previously created using ESO
+    // and not being used to get around the specified ID limitation
+    const canSpecifyID =
+      (options.overwrite && options.version) || SavedObjectsUtils.isRandomId(options.id);
+    if (options.id && !canSpecifyID) {
       throw new Error(
-        `Predefined IDs are not allowed for encrypted saved objects of type "${type}".`
+        'Predefined IDs are not allowed for saved objects with encrypted attributes, unless the ID has been generated using `SavedObjectsUtils.generateId`.'
       );
     }
```

Comment:
Nit: the same code is used in two different places, could be abstracted out to a function like
```ts
private throwIfIdIsNotAllowed(id: string | undefined, overwrite: boolean, version: string) {
  const canSpecifyID = (overwrite && version) || SavedObjectsUtils.isRandomId(id);
  if (id && !canSpecifyID) {
    throw new Error(
      'Predefined IDs are not allowed for saved objects with encrypted attributes, unless the ID has been generated using `SavedObjectsUtils.generateId`.'
    );
  }
}
```

Reply:
I have added a `getValidId` method that handles all that logic and also generates an id if none has been specified:
https://github.com/elastic/kibana/pull/84113/commits/459e0c4f25a7184c5c35da246369f62f8691eea7#diff-06a76df7e9f18756a4bcd62574dd9d6fd505a9ed7c8f275a1bdabd52a81a4bb6R298-R320